docs/design/values.md

```diff
@@ -0,0 +1,917 @@
+local `var` declaration.
+
+The next place where a Carbon expression requires an initializing expression is
+the expression operand to `return` statements. We expand more completely on how
+return statements interact with expressions, values, objects, and storage
+[below](#function-calls-and-returns).
+
+The last path that requires forming an initializing expression in Carbon is when
+attempting to convert a non-reference expression into an ephemeral reference
+expression, where temporary storage is materialized and provided to an
+initializing expression. The resulting object can then form the ephemeral
+reference expression.
+
+### Function calls and returns
+
+Functions with a return type and return statement in Carbon are called
+_returning functions_. Calls to these functions are modeled directly as
+initializing expressions -- they require storage as an input and when evaluated
+cause that storage to be initialized with an object. This means that when a
+function call is used to initialize some variable pattern as here:
+
+```carbon
+fn CreateMyObject() -> MyType {
+  return <return-expression>;
+}
+
+var x: MyType = CreateMyObject();
+```
+
+The `<return-expression>` in the `return` statement actually initializes the
+storage provided for `x`. There is no "copy" or other step.
+
+All `return` statement expressions are required to be initializing expressions
+and in fact initialize the storage provided to the returning function's call
+expression. This in turn causes the property to hold _transitively_ across an
+arbitrary number of function calls and returns. The storage is forwarded at each
+stage and initialized exactly once.
+
+#### Deferred initialization from values and references
+
+Carbon also makes the evaluation of returning function calls and return
+statements are tightly linked in order to enable more efficiency improvements.
+It allows the actual initialization performed by the `return` statement with its
+expression can be deferred from within the body of the function to the caller
+initializer expression if it can simply propagate a value or reference
+expression to the caller that is guaranteed to be alive and available to the
+caller.
+
+Consider the following code:
+
+```carbon
+fn SelectSecond(first: Point, second: Point, third: Point) -> Point {
+  return second;
+}
+
+fn UsePoint(p: Point);
+
+fn F(p1: Point, p2: Point) {
+  UsePoint(SelectSecond(p2, p1, p2));
+}
+```
+
+The call to `SelectSecond` must provide storage for a `Point` that can be
+initialized. However, Carbon allows an implementation of the actual
+`SelectSecond` function to not initialize this storage when it reaches
+`return second`. Because the expression `second` is a value expression which is
+necessarily valid in the caller, Carbon allows an implementation that simply
+communicates that this value expression is the return expression and _if
+necessary_ should initialize the temporary storage. This in turn allows the
```

Comment:
When you say "this value expression", which value expression do you mean? The phrasing of the sentence makes it sound like you mean the operand of `return` (i.e. `second`), but does that mean the same optimization could apply if the function ended with `return second.x;`? Or `return first + second;`? If so, that's way outside my mental model of function calling conventions, at least in a C-like language, so I feel like it would need a lot more explanation. Or would it be more accurate to say that "this value expression" actually refers to the argument expression that `second` is bound to (i.e. `p1`), and/or that we're actually talking about the _binding_ named `second`, rather than a specific expression that names it?

Reply:
(resolving as I think this is addressed)